pkg/auth/mtls_authhandler.go

```diff
@@ -0,0 +1,279 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+
+package auth
+
+import (
+	"context"
+	"crypto/tls"
+	"crypto/x509"
+	"errors"
+	"fmt"
+	"net"
+	"time"
+
+	"github.com/sirupsen/logrus"
+	"github.com/spf13/pflag"
+
+	"github.com/cilium/cilium/pkg/auth/certs"
+	"github.com/cilium/cilium/pkg/hive"
+	"github.com/cilium/cilium/pkg/hive/cell"
+	"github.com/cilium/cilium/pkg/identity"
+	"github.com/cilium/cilium/pkg/logging/logfields"
+	"github.com/cilium/cilium/pkg/policy"
+)
+
+type mtlsParams struct {
+	cell.In
+
+	CertificateProviders []certs.CertificateProvider `group:"certificateProviders"`
+}
+
+func newMTLSAuthHandler(lc hive.Lifecycle, cfg MTLSConfig, params mtlsParams, log logrus.FieldLogger) authHandlerResult {
+	if cfg.MTLSListenerPort == 0 {
+		log.Info("mTLS Auth Handler disabled is disabled as no port is configured")
+		return authHandlerResult{}
+	}
+	var cert certs.CertificateProvider
+	for _, c := range params.CertificateProviders {
+		if c == nil {
+			continue
+		}
+		if cert != nil {
+			log.Fatal("Multiple certificate providers configured, but only one is supported")
+		}
+		cert = c
+	}
+	if cert == nil {
+		log.Fatal("No certificate provider configured, but one is required")
+	}
+
+	mtls := &mtlsAuthHandler{
+		cfg:  cfg,
+		log:  log.WithField(logfields.LogSubsys, "mtls-auth-handler"),
+		cert: params.CertificateProviders[0],
+	}
+
+	lc.Append(hive.Hook{OnStart: mtls.onStart, OnStop: mtls.onStop})
+
+	return authHandlerResult{
+		AuthHandler: mtls,
+	}
+}
+
+type MTLSConfig struct {
+	MTLSListenerPort int `mapstructure:"mesh-auth-mtls-listener-port"`
+}
+
+func (cfg MTLSConfig) Flags(flags *pflag.FlagSet) {
+	flags.IntVar(&cfg.MTLSListenerPort, "mesh-auth-mtls-listener-port", 0, "Port on which the Cilium Agent will perfom mTLS handshakes between agents on")
```

Comment:
Nit: "Port on which the Cilium Agent will perfom mTLS handshakes between other Agents"